sw/device/silicon_creator/rom_ext/bootstrap.c

```diff
@@ -0,0 +1,102 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/rom_ext/bootstrap.h"
+
+#include <stdint.h>
+
+#include "sw/device/lib/base/hardened.h"
+#include "sw/device/silicon_creator/lib/base/chip.h"
+#include "sw/device/silicon_creator/lib/bootstrap.h"
+#include "sw/device/silicon_creator/lib/drivers/flash_ctrl.h"
+#include "sw/device/silicon_creator/lib/drivers/otp.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "flash_ctrl_regs.h"
+#include "hw/ip/otp_ctrl/data/otp_ctrl_regs.h"
+
+enum {
+  /*
+   * Maximum flash address, exclusive.
+   */
+  kMaxAddress =
+      FLASH_CTRL_PARAM_BYTES_PER_BANK * FLASH_CTRL_PARAM_REG_NUM_BANKS,
+  /*
+   * The total number of flash pages.
+   */
+  kNumPages = kMaxAddress / FLASH_CTRL_PARAM_BYTES_PER_PAGE,
+};
+
+rom_error_t bootstrap_chip_erase(void) {
+  // Unlike ROM bootstrap, we cannot erase entire banks at a time because that
+  // would erase the ROM_EXT region in the slot.
+  flash_ctrl_bank_erase_perms_set(kHardenedBoolTrue);
+  rom_error_t last_err = kErrorOk;
+  for (uint32_t i = 0; i < kNumPages; ++i) {
+    const uint32_t addr = i * FLASH_CTRL_PARAM_BYTES_PER_PAGE;
+    // Do not erase this page if it lies within ROM_EXT on either slot.
+    if (addr < CHIP_ROM_EXT_SIZE_MAX ||
+        (addr >= FLASH_CTRL_PARAM_BYTES_PER_BANK &&
+         addr < FLASH_CTRL_PARAM_BYTES_PER_BANK + CHIP_ROM_EXT_SIZE_MAX)) {
+      continue;
+    }
+    rom_error_t err = flash_ctrl_data_erase(addr, kFlashCtrlEraseTypePage);
+    if (err != kErrorOk) {
+      last_err = err;
+    }
+  }
+  flash_ctrl_bank_erase_perms_set(kHardenedBoolFalse);
+  HARDENED_RETURN_IF_ERROR(last_err);
+  return kErrorOk;
```

Comment:
```suggestion
  return last_err;
```

Reply:
Good catch. Don't want to create any gadgets :)